File: AI-LLM-Patterns/Tokenization/Byte-Pair-Encoding/bpe_tokenizer.cpp
#include <iostream>
#include <vector>
#include <string>
#include <map>
#include <sstream>
#include <algorithm>
#include <cassert>

using namespace std;

class BPETokenizer {
private:
    int numMerges;
    map<pair<string, string>, string> merges;

public:
    BPETokenizer(int mergesCount) : numMerges(mergesCount) {}

    void train(const string& text) {
        stringstream ss(text);
        string word;
        map<vector<string>, int> wordCounts;

        while (ss >> word) {
            vector<string> split;
            for (char c : word) {
                split.push_back(string(1, c));
            }
            split.push_back("</w>");
            wordCounts[split]++;
        }

        for (int i = 0; i < numMerges; i++) {
            map<pair<string, string>, int> pairs;

            for (auto const& [wordTuple, freq] : wordCounts) {
                if (wordTuple.size() < 2) continue;
                for (size_t j = 0; j < wordTuple.size() - 1; j++) {
                    pairs[{wordTuple[j], wordTuple[j + 1]}] += freq;
                }
            }

            if (pairs.empty()) break;

            pair<string, string> bestPair;
            int maxFreq = -1;
            for (auto const& [pair, freq] : pairs) {
                if (freq > maxFreq) {
                    maxFreq = freq;
                    bestPair = pair;
                }
            }

            if (maxFreq < 1) break;

            string newToken = bestPair.first + bestPair.second;
            merges[bestPair] = newToken;

            map<vector<string>, int> newWordCounts;
            for (auto const& [wordTuple, freq] : wordCounts) {
                vector<string> newWord;
                size_t j = 0;
                while (j < wordTuple.size()) {
                    if (j < wordTuple.size() - 1 && wordTuple[j] == bestPair.first && wordTuple[j + 1] == bestPair.second) {
                        newWord.push_back(newToken);
                        j += 2;
                    } else {
                        newWord.push_back(wordTuple[j]);
                        j++;
                    }
                }
                newWordCounts[newWord] = freq;
            }
            wordCounts = newWordCounts;
        }
    }

    vector<string> tokenize(const string& text) {
        stringstream ss(text);
        string word;
        vector<string> outputTokens;

        while (ss >> word) {
            vector<string> tokens;
            for (char c : word) {
                tokens.push_back(string(1, c));
            }
            tokens.push_back("</w>");

            while (tokens.size() > 1) {
                bool mergedAny = false;
                
                for (size_t i = 0; i < tokens.size() - 1; i++) {
                    auto checkPair = make_pair(tokens[i], tokens[i + 1]);
                    if (merges.find(checkPair) != merges.end()) {
                        string targetToken = merges[checkPair];
                        vector<string> newTokens;
                        size_t j = 0;
                        while (j < tokens.size()) {
                            if (j < tokens.size() - 1 && tokens[j] == checkPair.first && tokens[j + 1] == checkPair.second) {
                                newTokens.push_back(targetToken);
                                j += 2;
                            } else {
                                newTokens.push_back(tokens[j]);
                                j++;
                            }
                        }
                        tokens = newTokens;
                        mergedAny = true;
                        break;
                    }
                }
                if (!mergedAny) break;
            }
            for (const auto& token : tokens) {
                outputTokens.push_back(token);
            }
        }
        return outputTokens;
    }
};

int main() {
    BPETokenizer tokenizer(10);
    tokenizer.train("hug pug hug pug hug pug shug");
    vector<string> result = tokenizer.tokenize("hug pug");

    assert(result.size() < 8);
    assert(!result.empty());

    cout << "C++ BPE Tokenizer Test Passed!\n";
    return 0;
}

File: AI-LLM-Patterns/Tokenization/Byte-Pair-Encoding/bpeTokenizer.ts
export class BPETokenizer {
    private numMerges: number;
    private merges: Map<string, string> = new Map();

    constructor(numMerges: number) {
        this.numMerges = numMerges;
    }

    public train(text: string): void {
        const words = text.split(/\s+/);
        let wordCounts: Map<string[], number> = new Map();

        for (const word of words) {
            if (word.length === 0) continue;
            const split = [...word, "</w>"];
            
            // Check for existing matching array references
            let found = false;
            for (const [key, val] of wordCounts.entries()) {
                if (key.join(",") === split.join(",")) {
                    wordCounts.set(key, val + 1);
                    found = true;
                    break;
                }
            }
            if (!found) wordCounts.set(split, 1);
        }

        for (let i = 0; i < this.numMerges; i++) {
            const pairs: Map<string, number> = new Map();

            for (const [wordTuple, freq] of wordCounts.entries()) {
                for (let j = 0; j < wordTuple.length - 1; j++) {
                    const pairStr = `${wordTuple[j]},${wordTuple[j + 1]}`;
                    pairs.set(pairStr, (pairs.get(pairStr) || 0) + freq);
                }
            }

            if (pairs.size === 0) break;

            // Extract best pair
            let bestPairStr = "";
            let maxFreq = -1;
            for (const [pair, freq] of pairs.entries()) {
                if (freq > maxFreq) {
                    maxFreq = freq;
                    bestPairStr = pair;
                }
            }

            if (maxFreq < 1) break;

            const bestPair = bestPairStr.split(",");
            const newToken = bestPair.join("");
            this.merges.set(bestPairStr, newToken);

            const newWordCounts: Map<string[], number> = new Map();
            for (const [wordTuple, freq] of wordCounts.entries()) {
                const newWord: string[] = [];
                let j = 0;
                while (j < wordTuple.length) {
                    if (j < wordTuple.length - 1 && wordTuple[j] === bestPair[0] && wordTuple[j + 1] === bestPair[1]) {
                        newWord.push(newToken);
                        j += 2;
                    } else {
                        newWord.push(wordTuple[j]);
                        j++;
                    }
                }
                newWordCounts.set(newWord, freq);
            }
            wordCounts = newWordCounts;
        }
    }

    public tokenize(text: string): string[] {
        const words = text.split(/\s+/);
        const outputTokens: string[] = [];

        for (const word of words) {
            if (word.length === 0) continue;
            let tokens = [...word, "</w>"];

            while (tokens.length > 1) {
                let pairToMerge: string[] | null = null;
                
                // Scan tokens to see which of our rules applies first
                for (let i = 0; i < tokens.length - 1; i++) {
                    const checkKey = `${tokens[i]},${tokens[i + 1]}`;
                    if (this.merges.has(checkKey)) {
                        pairToMerge = [tokens[i], tokens[i + 1]];
                        break;
                    }
                }

                if (!pairToMerge) break;

                const matchKey = `${pairToMerge[0]},${pairToMerge[1]}`;
                const targetToken = this.merges.get(matchKey)!;
                const newTokens: string[] = [];
                let i = 0;
                while (i < tokens.length) {
                    if (i < tokens.length - 1 && tokens[i] === pairToMerge[0] && tokens[i + 1] === pairToMerge[1]) {
                        newTokens.push(targetToken);
                        i += 2;
                    } else {
                        newTokens.push(tokens[i]);
                        i++;
                    }
                }
                tokens = newTokens;
            }
            outputTokens.push(...tokens);
        }
        return outputTokens;
    }
}

// --- CI/CD Automated Test ---
const tokenizer = new BPETokenizer(10);
tokenizer.train("hug pug hug pug hug pug shug");
const result = tokenizer.tokenize("hug pug");

if (result.length < 8 && result.length > 0) {
    console.log("TypeScript BPE Tokenizer Test Passed!");
} else {
    process.exit(1);
}